src/pgw_s5s8_proxy.erl

```diff
@@ -734,31 +777,48 @@ build_context_request(#tunnel{remote = #fq_teid{teid = TEI}} = Tunnel, Bearer,
 msg(#tunnel{remote = #fq_teid{teid = RemoteCntlTEI}}, Type, RequestIEs) ->
     #gtp{version = v2, type = Type, tei = RemoteCntlTEI, ie = RequestIEs}.
 
-send_request(Tunnel, DstIP, DstPort, T3, N3, Msg) ->
-    gtp_context:send_request(Tunnel, DstIP, DstPort, T3, N3, Msg, undefined).
+send_request(Tunnel, DstIP, DstPort, T3, N3, Msg, ReqInfo) ->
+    gtp_context:send_request(Tunnel, DstIP, DstPort, T3, N3, Msg, ReqInfo).
 
-send_request(#tunnel{remote = #fq_teid{ip = RemoteCntlIP}} = Tunnel, T3, N3, Msg) ->
-    send_request(Tunnel, RemoteCntlIP, ?GTP2c_PORT, T3, N3, Msg).
+send_request(#tunnel{remote = #fq_teid{ip = RemoteCntlIP}} = Tunnel, T3, N3, Msg, ReqInfo) ->
+    send_request(Tunnel, RemoteCntlIP, ?GTP2c_PORT, T3, N3, Msg, ReqInfo).
 
-send_request(Tunnel, T3, N3, Type, RequestIEs) ->
-    send_request(Tunnel, T3, N3, msg(Tunnel, Type, RequestIEs)).
+send_request(Tunnel, T3, N3, Type, RequestIEs, ReqInfo) ->
+    send_request(Tunnel, T3, N3, msg(Tunnel, Type, RequestIEs), ReqInfo).
 
-initiate_session_teardown(sgw2pgw,
+initiate_session_teardown(sgw2pgw = Direction, From, connected,
 			  #{right_tunnel := Tunnel,
-			    proxy_context := #context{default_bearer_id = EBI}}) ->
+			    proxy_context := #context{default_bearer_id = EBI}} = Data) ->
     Type = delete_session_request,
     RequestIEs0 = [#v2_cause{v2_cause = network_failure},
 		   #v2_eps_bearer_id{eps_bearer_id = EBI}],
     RequestIEs = gtp_v2_c:build_recovery(Type, Tunnel, false, RequestIEs0),
-    send_request(Tunnel, ?T3, ?N3, Type, RequestIEs);
-initiate_session_teardown(pgw2sgw,
+    send_request(Tunnel, ?T3, ?N3, Type, RequestIEs, {Direction, From}),
+    maps:update_with(shutdown, [Direction|_], [Direction], Data);
+initiate_session_teardown(pgw2sgw = Direction, From, State,
 			  #{left_tunnel := Tunnel,
-			    context := #context{default_bearer_id = EBI}}) ->
+			    context := #context{default_bearer_id = EBI}} = Data)
+  when State == connected; State == connecting ->
     Type = delete_bearer_request,
     RequestIEs0 = [#v2_cause{v2_cause = reactivation_requested},
 		   #v2_eps_bearer_id{eps_bearer_id = EBI}],
     RequestIEs = gtp_v2_c:build_recovery(Type, Tunnel, false, RequestIEs0),
-    send_request(Tunnel, ?T3, ?N3, Type, RequestIEs).
+    send_request(Tunnel, ?T3, ?N3, Type, RequestIEs, {Direction, From}),
+    maps:update_with(shutdown, [Direction|_], [Direction], Data);
+initiate_session_teardown(_, _, _, Data) ->
+    Data.
+
+session_teardown_response(Answer, Direction, From, #{shutdown := Shutdown0} = Data) ->
+    case lists:delete(Direction, Shutdown0) of
+	[] ->
```

Comment:
Please, align the code block, it looks like used mix tabs and spaces.

Reply:
mix tab+space is indented